syncopy/datatype/discrete_data.py

```diff
@@ -506,6 +506,35 @@ def _get_unit(self, trials, units=None):
 
         return indices
 
+    @property
+    def waveform(self):
+        """The waveform of the spikes in the data.
+
+        This is a tiny part of the raw data around the time point at which a spike was detected
+        (by the recording hardware/software: keep in mind that the spike data Syncopy is working
+        with is already pre-processed). From this sequence of samples, one can derive
+        the waveform type of the spike (via classification), which may allow to
+        derive the type of neuron that produced the spike.
+        """
+        return self._waveform
+
+    @waveform.setter
+    def waveform(self, waveform):
+        """Set a waveform dataset from a numpy array, `None`, or an `h5py.Dataset` instance."""
+        if self.data is None:
+            if waveform is not None:
+                raise SPYValueError(f"non-empty SpikeData", "cannot assign `waveform` without data. " +
+                                    "Please assign data first")
+        if waveform is None:
+            self._set_dataset_property(waveform, 'waveform') # None
```

Comment:
It's a bit funny, with `.data` having shape `(1000, 3)` and trying to set a waveform with shape `(1000, 2)` this gets catched here with:
```python
>>> SPYValueError: Invalid value of `data`: 'data with shape (1000, 2, 10)'; expected dataset with shape (1000, 3, 10)
```
So probably the exception should say "Invalid value of `waveform` here"?! 
When trying with a waveform with shape `(100, 3)` I get:
```python
>>> SPYValueError: Invalid value of `waveform shape[0]=100 must equal nSpikes=1000. Please create one waveform per spike in data.`: expected wrong size waveform
```
Meaning that this dataset setter does not catch it, and hence you need the 2nd check below?

Reply:
Yes, that case is handled further down the line by existing machinery. I made the exception message more specific in be9082d.